spec/helpers.lua

```diff
@@ -80,9 +80,39 @@ local function unindent(str, concat_newlines)
   return (str:gsub("\n" .. prefix, repl):gsub("\n$", "")):gsub("\\r", "\r")
 end
 
+-- wrap conf_loader to disable existing environment variables that
+-- might otherwise interfere with the tests
+local function conf_loader(...)
+  local _getenv = os.getenv
+  os.getenv = function() end           -- luacheck: ignore
+  local conf, err = _conf_loader(...)
+  os.getenv = _getenv                  -- luacheck: ignore
+  return conf, err
+end
+
+-- generate shell command to 'unset' existing KONG_xxx variables
+local undo_environment = function()
+  local _, _, stdout, _ = pl_utils.executeex("export | grep KONG_")
+  if not stdout then
+    return ""
+  end
+
+  stdout = pl_utils.split(stdout, "\n")
+  local existing_vars = ""
+  for _, line in ipairs(stdout) do
+    local var = line:match("(KONG_.-)=")
+    if var then
+      existing_vars = ("%sunset %s; "):format(existing_vars, var)
```

Comment:
I don't think this is desirable. While the goal of avoiding ENV variables to interfere with the test suite is a *very* worthy one, that of having the test suite not interfere with the host's environment is at least as important.

Unfortunately, this will unset one's host KONG-related environment variable and replace them with testing values instead once `kong_exec()` runs...

I feel like we might need another approach. The same way, we cannot individually "stub" `os.getenv()` everywhere it is being used, that seems rather cumbersome and prompt to errors.

How about:

1. Mocking `os.getenv()` in `bin/busted`, so that no ENV variable from the host will interfere with the test suite, period.
2. Reviewing `kong_exec()` so that instead of feeding the Kong test instance some environment variables, we make it write to a temporary configuration file in the `servroot/` prefix, combined of the regular `spec/kong_tests.conf` and our overridden values?

I haven't given this too much thought, but these are worthy paths to explore I think?

Reply:
Another approach I believe I like even more: time and time again, I have found it useful to run a test suite with a different database with `KONG_DATABASE=... bin/busted ...`. Currently, our test suites are not all running on both databases, and this approach has made it possible. It is, I believe, just as beneficial to change other properties of the test instance from such environment variables.

I thus propose the following: the test instance does *not* consider `KONG_` prefixed env variables. Instead, it considers `KONG_TEST_` prefixed env variables.

A way we can do this is by introducing a new `test_instance = on|off` property in our configuration file (a hidden one, not documented in `kong.conf.default` and not visible in `:8001/`. When enabled, the `conf_loader` will load ENV variables prefixed with `KONG_TESTS_` instead of `KONG_`.

Thoughts?

Reply:
> Unfortunately, this will unset one's host KONG-related environment variable and replace them with testing values instead once `kong_exec()` runs...

Not sure what you mean here. They are only unset in the shell in which the kong command runs. They will not interfere with the host environment. Once the test command finishes, control is returned to the host shell, which still has all the original values.

Reply:
I like the `KONG_TEST_` approach. If you combine it with your initial proposal then we'd get an `os.getenv` in `bin/busted` (or maybe in `globalpatches` ?) that will return the value of `KONG_TEST_xxx` when we're looking up `KONG_xxx`, but only when testing.

Reply:
Hmmm I'm not sure what I was up to when I wrote that... 🤔 Yes, it should be fine and the spawned shell from `os.execute()` will do just fine with leaving the host environment untouched!

Reply:
If we can get away without patching `os.getenv` I'd much rather do that.

> A way we can do this is by introducing a new `test_instance = on|off` property in our configuration file (a hidden one, not documented in kong.conf.default and not visible in :8001/. When enabled, the conf_loader will load ENV variables prefixed with `KONG_TESTS_` instead of `KONG_`.

It seems to me this approach seems sufficient?

Reply:
Introducing that variable and modifying conf_loader means having to change code that also ends up in the final application. By patching it we would only modify behaviour at test time, and not introduce changes in the shipped application.

That was my reasoning, but of course patching is... well, patching

Reply:
That makes sense. But I'm afraid that patching `os.getenv` in `bin/busted` won't be sufficient, since we spawn new `kong` processes during tests.

An alternative would be to reset the actual environment variables at the top `bin/busted` script (and apply the `KONG_TEST_*` ones while we're at it).

I tried the approach here and it seems to work. Submitted PR #3075 with my tweak, check it out.